<a href="https://colab.research.google.com/github/hush-cz/ML_exp_tr_v1/blob/main/Prompt_Consistency_Experiment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Prompt Consistency Experiment

**Author:** Tomáš Vodáček
**Goal:** Ověřit, jak konzistentně odpovídá jazykový model (LLM) na faktické otázky při různém promptu.
**Method:** Použití open-source modelu z HuggingFace (DistilGPT-2).  
**Expected outcome:** Vyhodnocení, zda model odpovídá konzistentně, nebo vykazuje variabilitu.  
**Instrument:** Iphone and poor Malta wifi :)

In [ ]:
!pip install transformers

In [ ]:
from transformers import pipeline, set_seed

# inicializace generátoru textu
generator = pipeline("text-generation", model="distilgpt2")
set_seed(42)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/762 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/353M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Device set to use cpu


In [ ]:
prompts = [
    "What is the capital of France?",
    "France's capital city is ...",
    "Name the capital of France.",
    "Which city is the capital of France?",
    "Give me the capital of France."
]

In [ ]:
results = {}

for prompt in prompts:
    outputs = generator(prompt, max_length=20, num_return_sequences=3)
    results[prompt] = [out['generated_text'] for out in outputs]

results

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Both `max_new_tokens` (=256) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Both `max_new_tokens` (=256) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Both `max_new_tokens` (=256) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Setting `pad_token_id` to `eos_token_

{'What is the capital of France?': ['What is the capital of France?\n\nThis article is about the French capital. For other uses, see : French capital\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\nIt is not possible to have a French citizen in France if he is not a citizen of France. It is possible to have a French citizen in France if he is not a citizen of France. It is possible to have a French citizen in France if he is not a citizen of France. It is possible to have a French citizen in France if he is not a citizen of France.\nFrance has no national currency, and French currency is not issued by the French central bank. It is not issued by the French central bank.\nAs a result of the French Constitution, French citizens can purchase and purchase French currency in France. In the United States, citizens can purchase and purchase French currency in France.\nIn the United States, citizens can purchase and purchase French currency in France. For example, in the United States, the Fede

In [ ]:
for prompt, outputs in results.items():
    print(f"\nPrompt: {prompt}")
    for o in outputs:
        print("  ->", o.replace("\n"," "))


Prompt: What is the capital of France?
  -> What is the capital of France?  This article is about the French capital. For other uses, see : French capital                        It is not possible to have a French citizen in France if he is not a citizen of France. It is possible to have a French citizen in France if he is not a citizen of France. It is possible to have a French citizen in France if he is not a citizen of France. It is possible to have a French citizen in France if he is not a citizen of France. France has no national currency, and French currency is not issued by the French central bank. It is not issued by the French central bank. As a result of the French Constitution, French citizens can purchase and purchase French currency in France. In the United States, citizens can purchase and purchase French currency in France. In the United States, citizens can purchase and purchase French currency in France. For example, in the United States, the Federal Reserve is issuin

In [ ]:
correct_count = 0
total_count = 0

for prompt, outputs in results.items():
    for o in outputs:
        total_count += 1
        if "Paris" in o:
            correct_count += 1

accuracy = correct_count / total_count * 100
print(f"Consistency accuracy: {accuracy:.2f}% ({correct_count}/{total_count} answers contained 'Paris')")

Consistency accuracy: 13.33% (2/15 answers contained 'Paris')




## Conclusion

This simple experiment demonstrated that a small language model (DistilGPT-2)
was able to produce the correct answer ("Paris") only 2 out of 15 times
(~13% consistency) when asked the same fact-based question in different ways.

This highlights two important points:
1. Smaller open-source models often lack factual robustness.
2. Evaluating consistency across prompt variations is crucial for measuring reliability.

A natural next step would be to repeat this experiment with larger models
(e.g., GPT-Neo, LLaMA) and compare the consistency rates.

